core/ipc/grpc/server/src/main/java/org/opennms/core/ipc/grpc/server/OpennmsGrpcServer.java

```diff
@@ -0,0 +1,685 @@
+                    LOG.warn("RPC request from module: {} with RpcId:{} timedout ", responseHandler.getRpcModule().getId(),
+                            responseHandler.getRpcId());
+                    responseHandlerExecutor.execute(() -> responseHandler.sendResponse(null));
+                }
+            } catch (InterruptedException e) {
+                LOG.info("interrupted while waiting for an element from rpcTimeoutQueue", e);
+                Thread.currentThread().interrupt();
+                break;
+            } catch (Exception e) {
+                LOG.warn("error while sending response from timeout handler", e);
+            }
+        }
+    }
+
+
+    private void handleResponse(RpcResponseProto responseProto) {
+
+        if (Strings.isNullOrEmpty(responseProto.getRpcId())) {
+            return;
+        }
+        // Handle response from the Minion.
+        RpcResponseHandler responseHandler = rpcResponseMap.get(responseProto.getRpcId());
+        if (responseHandler != null && responseProto.getRpcContent() != null) {
+            responseHandler.sendResponse(responseProto.getRpcContent().toStringUtf8());
+        } else {
+            LOG.debug("Received a response for request for module: {} with RpcId:{}, but no outstanding request was found with this id." +
+                    "The request may have timed out", responseProto.getModuleId(), responseProto.getRpcId());
+        }
+    }
+
+    private boolean sendRequest(RpcRequestProto requestProto) {
+        StreamObserver<RpcRequestProto> rpcHandler = getRpcHandler(requestProto.getLocation(), requestProto.getSystemId());
+        if (rpcHandler == null) {
+            LOG.warn("No RPC handlers found for location {}", requestProto.getLocation());
+            return false;
+        }
+        try {
+            sendRpcRequest(rpcHandler, requestProto);
+            return true;
+        } catch (Throwable e) {
+            LOG.error("Encountered exception while sending request {}", requestProto, e);
+        }
+        return false;
+    }
+
+    /**
+     * Writing message through stream observer is not thread safe.
+     */
+    private synchronized void sendRpcRequest(StreamObserver<RpcRequestProto> rpcHandler, RpcRequestProto rpcMessage) {
+        rpcHandler.onNext(rpcMessage);
+    }
+
+    private synchronized StreamObserver<RpcRequestProto> getRpcHandler(String location, String systemId) {
+
+        if (!Strings.isNullOrEmpty(systemId)) {
+            return rpcHandlerByMinionId.get(systemId);
+        }
+        Collection<StreamObserver<RpcRequestProto>> streamObservers = rpcHandlerByLocation.get(location);
+        if (streamObservers.isEmpty()) {
+            return null;
+        }
+        Iterator<StreamObserver<RpcRequestProto>> iterator = rpcHandlerIteratorMap.get(streamObservers);
+        if (iterator == null) {
+            iterator = Iterables.cycle(streamObservers).iterator();
+            rpcHandlerIteratorMap.put(streamObservers, iterator);
```

Comment:
How does this iterator behave when the underlying collection changes? Should we clear the entry from the map when new handlers get registered?

Reply:
Handled this differently with location as key and updating iterator in add/remove handler methods.